### ISSUE TITLE
Add compiler warning when void delagate returns Task

### ISSUE THREAD

**Issue by sungam3r** [dotnet/roslyn#58076]
**Brief description:**

I would like to see some warning when code returns `Task`/`ValueTask` (or other awaitable object) from `Action`-like delegate. There is existing `CS4014` warning about not awaited call. New warning is not about async code at all (generally) but about not used _return result_ from `void` method. See [example](https://dotnetfiddle.net/tY0jUI). I ran into a hardly understandable error when updating one of the libraries. One of the library methods has received new overload and after recompiling the caller become (unnoticed) to use another overload. As a result, the code stopped awaiting `Task` object. This leads to unpredictable consequences in racing conditions.

**Languages applicable:**

C#, maybe VB also.

**Code example that the analyzer should report:**

See [example](https://dotnetfiddle.net/tY0jUI).

**Comment by sungam3r** [comment]
Copied example:

```c#
public class Program
{
    static async Task Main(string[] args)
    {
        object someVariable = null;

        Func<Task> func = async () =>
        {
            await Task.Delay(1000);
            someVariable = new object();
        };


        // No warning about Task -> Action closure on DoSomething V2
        DoSomething(() => func());

        // Works well for V1, fails for V2 with NullReferenceException
        Console.WriteLine(someVariable.GetType());


        // Just to demonstrate compiler warning.
        func(); // Warning CS4014  Because this call is not awaited, execution of the current method
                // continues before the call is completed.Consider applying the 'await' operator to
                // the result of the call.
    }

    // DoSomething from some library code V1
    // Comment this method to emulate V1 -> V2 upgrade
    static void DoSomething(Func<Task> callback)
    {
        callback().Wait();
    }

    // DoSomething from some library code V2
    static void DoSomething(Action callback)
    {
        callback();
    }
}
```

**Comment by CyrusNajmabadi** [comment]
If you is a request for an analyzer, that should go to roslyn-analyzers

**Comment by sharwell** [comment]
This is already available in Microsoft.VisualStudio.Threading.Analyzers (and can be used in projects unrelated to Visual Studio):

https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD100.md
https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD101.md

**Comment by sungam3r** [comment]
Not applicable. I installed these analyzers and see nothing new. These analyzers are for async code. The code in example is sync:
> DoSomething(() => func());

**Comment by sungam3r** [comment]
@CyrusNajmabadi Yes, I would like to . @sharwell Please transfer this issue if you can.

**Comment by sharwell** [comment]
> Not applicable. I installed these analyzers and see nothing new. These analyzers are for async code. The code in example is sync:
> 
> > DoSomething(() => func());

Ah, I see what you mean. This scenario is intended to be covered by [VSTHRD110](https://github.com/microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD110.md) (and it would be a bug if VSTHRD110 is not reported in this case).

**Comment by sungam3r** [comment]
> and it would be a bug if VSTHRD110 is not reported in this case

Obviously, it is.
